### ISSUE TITLE
Update CKEditor 5 to v44.1.0 for Backdrop 1.30.0

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#6695]
Continuing from https://github.com/backdrop/backdrop-issues/issues/6481, this issue is to track updating to the latest CKEditor 5 version.

Originally filed by @klonos:

> It would be nice to be updating to the latest release regularly (I am proposing with each minor release - we could do it in the 2 weeks between feature freeze and final release). ...if we don't keep up and we are several versions behind, then if a security release comes out we risk rushing to release the secure version, which might be introducing breaking changes, and we won't have enough time to test/adjust/fix things.

We are currently shipping core with v5, build ~42.0.2~ v43.2.0 (updated Oct 27, 2024).

https://ckeditor.com/blog/categories/releases highlights the main improvements with new releases:

- CKEditor v43.0.0 - [All-new Merge Fields and Export to Word v2](https://ckeditor.com/blog/ckeditor-43-0-0-release-highlights/)
- CKEditor v43.1.0 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v43.1.0)
- CKEditor v43.1.1 [related SA](https://github.com/ckeditor/ckeditor5/security/advisories/GHSA-rgg8-g5x8-wr9v)
- CKEditor v43.2.0 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v43.2.0)
- CKEditor v43.3.0 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v43.3.0)
- CKEditor v43.3.1 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v43.3.1)
- CKEditor v44.0.0 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v44.0.0)
- CKEditor v44.1.0 [release notes](https://github.com/ckeditor/ckeditor5/releases/tag/v44.1.0)

More versions will probably be released before we do the next update.

## More resources related to releases

- https://ckeditor.com/docs/ckeditor5/latest/updating/guides/changelog.html has more detailed changelogs.
- https://github.com/ckeditor/ckeditor5/releases lists bug fix releases.

## Core functionality checklist

- [ ] Link dialog and balloon toolbar
- [ ] Image dialog (open it, upload or select image and insert, replace image)
- [ ] Image caption toggle
- [ ] Image alignment
- [ ] Image drag-and-drop upload
- [ ] Styles configuration and usage in content
- [ ] Try different editor toolbar setups (less toolbar buttons, more toolbar buttons, different button combinations)

**Comment by indigoxela** [comment]
@quicksketch should I create an individual issue per version, or is it OK to attach PRs here (until we close it as confusing). :wink: We should put our checklist somewhere, so maybe an issue per version makes sense?

Currently version **43.1.0** is the latest. Big relief: they didn't touch the DLL (build) concept, yet, so our approach still works. Even without changing anything besides the library.
https://github.com/backdrop/backdrop/pull/4875

**Comment by quicksketch** [comment]
Let's use this issue until we manage to merge a CKEditor version into core, then we'll relabel this issue "Update CKEditor to version xx.yy.zz". Then we'll make another issue to track updating to the next latest version again.

> Even without changing anything besides the library.

That's great! If it's truly compatible, we might be able to put it into a minor release? i.e. 1.29.1

**Comment by indigoxela** [comment]
> If it's truly compatible, we might be able to put it into a minor release? i.e. 1.29.1

I belief, @klonos suggested to update CKE in minor releases - unless security related, which doesn't seem to apply to v43.1.0 or v43.0.0.

We already win, if we're not falling too far behind, and if we have tested in-between versions (merged or not), so a security update doesn't force us to rush in huge (unknown) changes. There are usually hundreds of commits per CKE minor release. Impossible for us to determine, which ones are relevant for us.

**Comment by indigoxela** [comment]
Time to update CKEditor again. Here's now 43.2.0 and it does seem to work just fine - needs more thorough testing, though, as usual.

There has been a security advisory for v43.1.1 (the previous version), but note that this is a "theoretical and unexploitable" issue, and furthermore is irrelevant for Backdrop, as we're not using the block toolbar, anyway. I belief, we can ignore that, but some confirmation would be nice.

**Comment by quicksketch** [comment]
I checked and we do actually have the `BlockToolbar` plugin bundled into our `ckeditor5-dll.js` file, however it's not enabled out of the box and it's not even possible to enable via the UI or through config, you would have had to enable it via custom code. Unless it's possible to exploit, an isolated bug that cannot be reached does not warrant a security release (or even an security announcement). It would be good to update none-the-less. As a general statement (not specific to @indigoxela), if it is found that it is possible to exploit please write an email to security@backdropcms.org, see [reporting a security vulnerability documentation](https://backdropcms.org/reporting-security-issues).

**Comment by indigoxela** [comment]
Rebased, a fresh sandbox is ready for testing. :wink:

**Comment by quicksketch** [comment]
I ran through the [CKEditor testing checklist](https://github.com/backdrop/backdrop-issues/wiki/CKEditor:-Checklist-for-manual-tests) plus other manual testing. Everything seems to be working as expected; all the Backdrop-CKEditor integration points seem to be working as expected.

**Comment by quicksketch** [comment]
Leaving this open if we need to update again in the 1.30.0 release. As of right now, we're fully up-to-date with CKEditor 5.

**Comment by indigoxela** [comment]
Their release cycles are still very short. We'll see, at which version number we are for 1.30.0 :grin: 

V43.3.0 just came out two days ago. List in issue description complemented.

**Comment by indigoxela** [comment]
V43.3.1 is the most recent one now.

Here's a [PR](https://github.com/backdrop/backdrop/pull/4908) for it. The editor still seems to works fine (DLL build).

**Comment by quicksketch** [comment]
From https://github.com/ckeditor/ckeditor5/releases/tag/v43.1.0:

> Reverted `config.sanitizeHtml`. In v43.0.0 we made a decision to move `config.htmlEmbed.sanitizeHtml` to a top-level property `config.sanitizeHtml`. However, we realized that it was a wrong decision to expose such a sensitive property in a top-level configuration property. Starting with v43.1.0 you should again use `config.htmlEmbed.sanitizeHtml` and/or `config.mergeFields.sanitizeHtml`. The editor will throw an error if `config.sanitizeHtml` is used. See the [migration guide](https://ckeditor.com/docs/ckeditor5/latest/updating/guides/update-to-43.html#reverted-recently-introduced-configsanitizehtml) for additional context behind this decision.

API-change whiplash. I think our practices of updating CKEditor for minor-releases only is a good one.

**Comment by indigoxela** [comment]
While rebasing, I realized, we're already behind again... :disappointed: ... v44 has been released yesterday, breaking changes...

I wonder, if we should wait for at least 44.0.1 (first bugfix release after major upgrade). Just in case, they need to revert something (again).

**Comment by indigoxela** [comment]
We now have two options available:

- v44.1.0 - the latest in https://github.com/backdrop/backdrop/pull/4946
- v43.3.1 - from November, in https://github.com/backdrop/backdrop/pull/4908

Both seem to work but need more thorough testing. And a decision, which one we want for 1.30.0.

For Both a fresh sandbox is available for testing. What to test? See the checklist in this issue's description.

**Comment by quicksketch** [comment]
Looks like your plan to wait for an additional release worked. :)

If it's feasible, going to the latest available version would be preferred. I'll take a look at the `v44.1.0` PR tomorrow.

**Comment by olafgrabienski** [comment]
I've had a look at the [PR for version v44.1.0](https://github.com/backdrop/backdrop/pull/4946). If I didn't miss anything, all the things from the checklist work like a charm:

- [x] Link dialog and balloon toolbar
- [x] Image dialog (open it, upload or select image and insert, replace image)
- [x] Image caption toggle
- [x] Image alignment
- [x] Image drag-and-drop upload
- [x] Styles configuration and usage in content
- [x] Try different editor toolbar setups (less toolbar buttons, more toolbar buttons, different button combinations)

Note: Apart from Basic, I tried _only one_ different toolbar setup (where I used nearly all buttons). So, some more testing with different toolbar setups is welcome.
